docs/compiler_passes.rst

```diff
@@ -1,11 +1,150 @@
 Compiler Passes
 ===============
 
```

Comment:
This file could really benefit from a figure. 

I am thinking similar to the following (from projectQ) that shows the different stages and possible entry points. 
![Screenshot 2019-12-10 at 20 25 22](https://user-images.githubusercontent.com/6142932/70561698-344bac00-1b8b-11ea-8674-98d1b18a2893.png)